{
 "metadata": {
  "name": "",
  "signature": "sha256:d5345fb7e2ff879996a422daecc800bbb9d99c0c328f79abfcb216b87dcde8b8"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Kraken is an algorithm that inspects sequence reads from a sample presumably containing multiple species and attempts to identify the species present ([Wood & Salzberg 2014](genomebiology.com/2014/15/3/R46)). This is most useful for microbiologists trying to identify bacteria species in environments such as a body of water or an animal's gut. This field of study is known as *metagenomics*.\n",
      "\n",
      "Kraken stores a database of all $k$-mers from the 2,256 complete reference genomes contained in [the NCBI's RefSeq database](http://www.ncbi.nlm.nih.gov/refseq/). The reference genomes' associated species are organized in a *phylogenetic tree* in which each species' node is attached to that of its most immediate ancestor.\n",
      "\n",
      "The production code uses a lot of optimization techniques. After all, the novelty of Kraken is not the problem it solved (many metagenomic classification algorithms already existed), but rather its speed and memory efficiency. However, our Python implementation will ignore most incremental optimizations for the sake of readability and concision, while using the same general data formats as the production code.\n",
      "\n",
      "We will populate our phylogenetic tree using data files supplied for Minikraken, which uses all of the reference genomes supplied by RefSeq but only a small subset of their $k$-mers."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from pprint import pprint\n",
      "\n",
      "nodes = \"/Volumes/8 GB FLASH/minikraken_20140330/taxonomy/nodes.dmp\"\n",
      "names = \"/Volumes/8 GB FLASH/minikraken_20140330/taxonomy/names.dmp\"\n",
      "\n",
      "\n",
      "class PhyloNode:\n",
      "    \n",
      "    def __init__(self, nodeID, parentID, taxonType):\n",
      "        self.nodeID = nodeID\n",
      "        self.parentID = parentID\n",
      "        self.name = None\n",
      "        self.taxonType = taxonType\n",
      "        self.children = set()\n",
      "        self.parent = None\n",
      "\n",
      "    def getLCA(descendants):\n",
      "        if not descendants:\n",
      "            return self\n",
      "        if descendants.issubset(self.children):\n",
      "            return self\n",
      "        possibleDescendants = descendants.intersect(self.children)\n",
      "        for child in children:\n",
      "            childsNondescendants = child.nonchildrenIn(descendants)\n",
      "            if not childsNondescendants:\n",
      "                return child.getLCA(descendants)\n",
      "            else:\n",
      "                possibleDescendants.intersection_update(\n",
      "                \n",
      "        elif not set.intersect([child.nonchildrenIn(descendants.intersect(self.children)) for child in self.children]):\n",
      "            return True\n",
      "        else:\n",
      "            return False\n",
      "        \n",
      "    # determines which nodeIDs in the set are not children of this node (inclusive)\n",
      "    def nonchildrenIn(nodeIDs):\n",
      "        if not nodeIDs:\n",
      "            return set()\n",
      "        else:\n",
      "            if nodeIDs.contains(self.nodeID):\n",
      "                nodeIDs.remove(self.nodeID)\n",
      "            return set.intersection([child.childrenIn(nodeIDs) for child in self.children])\n",
      "\n",
      "\n",
      "class PhyloTree:\n",
      "    \n",
      "    def __init__(self, nodesFileName, namesFileName):\n",
      "        self.size = 0\n",
      "        # used to quickly index node by nodeID\n",
      "        self.nodes = []\n",
      "\n",
      "        namesFile = open(namesFileName, \"r\")\n",
      "        nodesFile = open(nodesFileName, \"r\")\n",
      "\n",
      "        # populate nodes\n",
      "        for line in nodesFile:\n",
      "            nodeID, parentID, taxonType, _ = map(lambda x: x.strip(), line.split('|', 3))\n",
      "            nodeID, parentID = int(nodeID), int(parentID)\n",
      "            #pad for missing nodes\n",
      "            while nodeID > len(self.nodes):\n",
      "                self.nodes.append(None)\n",
      "            self.nodes.append(PhyloNode(nodeID, parentID, taxonType))\n",
      "\n",
      "        # populate children lists\n",
      "        for node in self.nodes:\n",
      "            if node and node.parentID:\n",
      "                node.parent = self.nodes[node.parentID]\n",
      "                node.parent.children.add(node)\n",
      "\n",
      "        # add names to nodes\n",
      "        for line in namesFile:\n",
      "            nodeID, name, _, nameType, _ = map(lambda x: x.strip(), line.split('|'))\n",
      "            nodeID = int(nodeID)\n",
      "            if nameType == \"scientific name\":\n",
      "                self.nodes[nodeID].name = name\n",
      "        \n",
      "\n",
      "\n",
      "phyloTree = PhyloTree(nodes, names)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}